design/README.md

```diff
@@ -0,0 +1,285 @@
+
+### App Publishing
+
+April wants to launch an app on Fission's service. She goes to the Fission CLI, and signs up for the service with her email, and registers the username `april2023`. She's prompted to set up a Passkey (which will be her root DID). A WNFS is created, delegates a superuser UCAN to `did:key:zAprilPasskey`. April is prompted to register an app name or use the default (autogenerated `friendly-pink-dragon`), and select a directory that she wants to upload (`./`). The app directory is uploaded via CAR Mirror, and DNSLink at `friendly-pink-dragon.fission.app` is set.
+
+### Managed Homestar FaaS
+
+Terri wants to register a function on Odd Functions. They go to the Fission CLI, and signs up for the service with their email, and registers the username `terri2023`. They're prompted to set up a Passkey (which will be their root DID). The Fission FaaS service then delegates 1000 credits to `did:key:zTerriPasskey`.
+
+# Approach
+
+## Event Sourcing
+
+Instead of recording the `createdAt` and latest `modifiedAt`, the Fission server records immutable events. This stores more data in the database, but has multiple advantages:
+
+- Schema migrations can be reconstituted performed directly or by playing over the log
+- Post hoc analysis and debugging is possible
+- Nondestructive updates makes fixing bugs in-place significantly easier
+
+## Resource DIDs
+
+In the previous (Haskell) server, the user's DID was registered directly in the (central) database. This approach has a few drawbacks:
+
+- Requires a DB lookup to find the associated owner for an account (or other resource)
+- Doesn't provide a solution to uniquely identifying a single resource in a one-to-many relationship
+- Has ACL semantics at the final step of the delegation (which is not unheard of in capability modeling)
+
+Instead, if every resource (include the account itself) is uniquely identified by a DID, then UCANs can be used throughout without resorting to an external ownership list. With this approach, the keys for the resource may be stored in a KMS, or ideally _dropped completely_ after delegating root access to one or more controllers. To solve for recovery and a server-controlled revocation fail-safe, a key may be kept offline (i.e. a "cold DID") and delegated to.
+
+Here is one example of such a DID hierarchy:
+
+``` mermaid
+flowchart
+    RootServerDID --> 2023ServerDID
+    2023ServerDID --> Jan2023ServerDID
+    2023ServerDID --> Feb2023ServerDID
+    2023ServerDID --> Mar2023ServerDID
+
+    RootServerDID --> 2024ServerDID
+    2024ServerDID --> Jan2024ServerDID
+    2024ServerDID --> Feb2024ServerDID
+    2024ServerDID --> Mar2024ServerDID
+
+    RootServerDID --> ...
+```
+
+Here is one such flow, with a cold server DID and monthly hot server DID:
```

Comment:
Is the assumption that the "cold server DID" will never be rotated?

Reply:
Yeah that's the idea. You _can_ rotate it, but it's basically forking the backend. It's a very powerful key in the system, so it should never be made hot (sign stuff out of band manually, very high salience, BLS multisig something something).

![](https://media.tenor.com/AiLDjTytWaMAAAAd/goldeneye-goldeneye-key.gif)